File: 2170-count-number-of-maximum-bitwise-or-subsets/count-number-of-maximum-bitwise-or-subsets.cpp
class Solution {
public:
    int countMaxOrSubsets(vector<int>& nums) {


        int best = 0;
        int freq = 0; 

        for (int subset = 0; subset < pow(2, nums.size()); subset += 1){

            int curr = 0; 

            for (int i = 0; i < nums.size(); i++){
                if ((subset >> i) & 1){
                    curr = curr | nums[i];
                }
            }

            if (curr == best){
                freq += 1; 
            }
            else if (curr > best){
                freq = 1; 
                best = curr; 
            }

        }

        return freq; 
    }
};

File: 2170-count-number-of-maximum-bitwise-or-subsets/count-number-of-maximum-bitwise-or-subsets.py
"""

This feels like brute force simply
Go thru all subsets, get bitwise or

"""

class Solution:
    def countMaxOrSubsets(self, nums: List[int]) -> int:

        best = 0
        freq = 0

        def dfs(i, curr):
            nonlocal best
            nonlocal freq

            if i == len(nums):
                if curr > best: 
                    freq = 1
                    best = curr
                elif curr == best: 
                    freq += 1
                return

            # dont chooose i
            dfs(i + 1, curr)
            # choose i
            dfs(i + 1, curr | nums[i])
        
        dfs(0, 0)
        return freq
            

        